task/mgmt-gateway/src/main.rs

```diff
@@ -0,0 +1,466 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+#![no_main]
+
+use drv_stm32h7_usart::Usart;
+use gateway_messages::sp_impl;
+use gateway_messages::sp_impl::Error as MgsDispatchError;
+use gateway_messages::sp_impl::SerialConsolePacketizer;
+use gateway_messages::sp_impl::SocketAddrV6;
+use gateway_messages::IgnitionCommand;
+use gateway_messages::Request;
+use gateway_messages::SerialConsole;
+use gateway_messages::SerializedSize;
+use gateway_messages::SpComponent;
+use gateway_messages::SpMessage;
+use gateway_messages::SpMessageKind;
+use gateway_messages::SpPort;
+use ringbuf::ringbuf;
+use ringbuf::ringbuf_entry;
+use task_net_api::Address;
+use task_net_api::Net;
+use task_net_api::NetError;
+use task_net_api::SocketName;
+use task_net_api::UdpMetadata;
+use tinyvec::ArrayVec;
+use userlib::sys_get_timer;
+use userlib::sys_irq_control;
+use userlib::sys_recv_closed;
+use userlib::sys_set_timer;
+use userlib::task_slot;
+use userlib::TaskId;
+use userlib::UnwrapLite;
+
+mod mgs_handler;
+
+use self::mgs_handler::MgsHandler;
+
+task_slot!(NET, net);
+task_slot!(SYS, sys);
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum Log {
+    Empty,
+    Wake(u32),
+    Rx(UdpMetadata),
+    DispatchError(MgsDispatchError),
+    SendError(NetError),
+    MgsMessage(MgsMessage),
+    UsartTx { num_bytes: usize },
+    UsartTxFull { remaining: usize },
+    UsartRx { num_bytes: usize },
+    UsartRxOverrun,
+    SerialConsoleSend { len: u16 },
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum MgsMessage {
+    Discovery,
+    IgnitionState {
+        target: u8,
+    },
+    BulkIgnitionState,
+    IgnitionCommand {
+        target: u8,
+        command: IgnitionCommand,
+    },
+    SpState,
+    SerialConsoleWrite {
+        length: u16,
+    },
+}
+
+ringbuf!(Log, 16, Log::Empty);
+
+// Must match app.toml!
+const NET_IRQ: u32 = 1;
+const USART_IRQ: u32 = 2;
+
+// Must not conflict with IRQs above!
+const TIMER_IRQ: u32 = 4;
+
+// Send any buffered serial console data to MGS when our oldest buffered byte is
+// this old, even if our buffer isn't full yet.
+const SERIAL_CONSOLE_FLUSH_TIMEOUT_MILLIS: u64 = 500;
+
+const SOCKET: SocketName = SocketName::mgmt_gateway;
+
+#[export_name = "main"]
+fn main() {
+    let mut usart_handler = UsartHandler::new(configure_usart());
+    let mut net_handler = NetHandler::default();
+
+    // Enbale USART interrupts.
+    sys_irq_control(USART_IRQ, true);
+
+    let mut note = NET_IRQ;
+    loop {
```

Comment:
I was worried about the net task having already received a packet before we called `recv()` and it therefore not knowing it needed to notify us, but after writing that out, it's not possible, is it? (Since the fact that we're the socket owner is known statically.)

Reply:
Fixed in 5675738f